bk-nodes/bk_db.py

```diff
@@ -0,0 +1,241 @@
+
+        self.cur=self.db.cursor()
+        return
+
+
+    def nodes_log_add(self, logData):
+        #node_id, table_name, operation, field_name, new_value, source, effective_time=None
+        fields = '`node_id`, `table_name`, `operation`, `field_name`, `new_value`, `source`, `effective_time`'
+        values_s = '%s, %s, %s, %s, %s, %s, %s'
+        #values = [node_id, table_name, operation, field_name, new_value, source]
+        
+        smallest_timestamp = None 
+        values = []
+        for log in logData:
+            effective_time =  dateutil.parser.parse(log["effective_time"])
+            
+            if (not smallest_timestamp) or (effective_time < smallest_timestamp):
+                smallest_timestamp = effective_time
+
+            value_tuple =(log["node_id"], log["table_name"], log["operation"], log["field_name"], log["new_value"], log["source"], effective_time)
+            values.append(value_tuple)
+
+
+
+
+        self.cur.execute("LOCK TABLES `nodes_log` WRITE")
+
+        stmt = f'INSERT INTO `nodes_log` ( {fields} ) VALUES ({values_s})'
+        #debug_stmt = stmt
+        for tup in values:
+            debug_stmt = stmt
+            for i in tup:
+                debug_stmt = debug_stmt.replace("%s", f'"{i}"', 1)
+            print(f'debug_stmt: {debug_stmt}', flush=True)
+
+        #print(f'debug_stmt: {debug_stmt}', file=sys.stderr)
+        
+        self.cur.executemany(stmt, values )
+        #self.cur.execute("UNLOCK TABLES")
+        self.db.commit()
+
+        try:
+
+            self.replay_log(replay_from_timestamp = smallest_timestamp)
+        except Exception as e:
+            raise Exception("Log replay failed: "+str(e))
+
+        return
+
+    
+    # a) replay log from scratch/zero
+    # b) replay log from effective_time (as changes)
+    # replay has to merge log entries with same timestamp !
+    def replay_log(self, replay_from_timestamp=None):
+
+       
+        #fields = ['node_id', 'table_name', 'operation', 'field_name', 'new_value', 'source', 'effective_time', 'modified_time']
+        #fieldIndex = {}
+        #for f in range(len(fields)):
+        #    fieldIndex[fields[f]] = f
+        fields = table_fields['nodes_log']
+        fieldIndex = table_fields_index['nodes_log']
+
+        fields_str = ', '.join(fields)
+
+        stmt = 'LOCK TABLES `nodes_history` WRITE , `nodes_log` READ'
+       #self.cur.execute(stmt)
+       # stmt = 'LOCK TABLES `nodes_log` READ'
+        self.cur.execute(stmt)
+        #self.db.commit()
+
+        if replay_from_timestamp:
+            # delete all entries newer than replay_from_timestamp in history
+
+            # microseconds in the timestamp lead to wrong behaviour, because microseconds are not in mysql it seems 
+            replay_from_timestamp = replay_from_timestamp.replace(microsecond=0)
+
+            stmt = f'DELETE FROM `nodes_history` WHERE `timestamp` >= %s'
+            print(f'statement: {stmt}', flush=True)
```

Comment:
sry, I can't follow you

Reply:
The print statement seen in the logs looks like this

```
statement: DELETE FROM `nodes_history` WHERE `timestamp` >= %s
```

It would be nice if it had the `%s` replaced by the actual value.

You did this with the `SELECT` statements on line 137

```
debug_stmt: SELECT id, node_id, table_name, operation, field_name, new_value, source, effective_time, modified_time FROM `nodes_log` WHERE `effective_time` >= "2021-01-17 18:46:31" ORDER BY `effective_time`
total_count: 4
```

Reply:
Oh, you are referring to the debug message.... Added that.